### ISSUE TITLE
disallowed_types for enum variants

### ISSUE THREAD

**Issue by DusterTheFirst** [rust-lang/rust-clippy#10154]
### What it does

Acts like the `disallowed_*` family of lints but would prevent the construction of a specific enum variant.

### Lint Name

disallowed_variants

### Category

style

### Advantage

Disallows usage of a particular enum variant in your code that may be unsupported in the context you are writing in, but may make sense for a library to expose for use.

My use case is with prometheus and openmetrics. prometheus v1 can be seen as a subset of openmetrics (to an extent) and so using an openmetrics compatible library with lint restrictions on the allowed types and enum variants would prevent the programmer from making the mistake of including a non-prometheus feature in their code.

### Drawbacks

As this does not exist yet, and has no mention from what I have seen, this lint may not be widely used.

### Example

In a context where unknown is not a valid metric type.

```rust
  const TYPE: MetricType = MetricType::Unknown;
```

Could be written as:

```rust
  const TYPE: MetricType = MetricType::Gauge;
```
